File: ocr-data-pipeline/src/main/scala/com/greenleaf/lor/ocr/pipeline/KeyParser.scala
package com.greenleaf.lor.ocr.pipeline

import java.io.FileReader
import java.text.SimpleDateFormat
import java.util.Date

import com.greenleaf.lor.ocr.pipeline.model.{LORMetaData, LORWriter, UserMetaData}
import com.typesafe.scalalogging.StrictLogging
import org.apache.commons.csv.{CSVFormat, CSVParser, CSVRecord}

import scala.util.{Failure, Success, Try}
import scala.jdk.CollectionConverters._

object KeyParser extends StrictLogging {
  val dateFormatter = new SimpleDateFormat("DD/MM/YYYY")

  def parseKey (fileName: String): Seq[UserMetaData] = {
    val source = new FileReader(fileName)
    val parser = new CSVParser(source, CSVFormat.DEFAULT.withHeader())

    Try (
      parser.asScala.foldLeft(Map[Int, UserMetaData]())((metaData, record) => {
        val id = record.get(0).toInt
        val existingMetaData = metaData.get(id)
        val userData = existingMetaData match {
          case Some(d) => d
          case None => extractUserData(record)
        }
        val LORData = extractLORData(record)
        metaData.updated(
          id,
          userData.copy(
            lorMetaData = (userData.lorMetaData :+ LORData)
          )
        )
      })
    ) match {
      case Success(userMetaData) =>
        parser.close()
        source.close()
        userMetaData.values.toSeq
      case Failure(ex) =>
        parser.close()
        source.close()
        throw new Exception(s"Failed to parse key file: ${fileName}", ex)
    }
  }

  private def extractUserData (record: CSVRecord) =
    UserMetaData (
      extractByHeader(record, "Participant Number").toInt,
      extractByHeader(record, "Medical School"),
      extractByHeader(record, "Gender"),
      extractByHeader(record, "Race (Self Identification)"),
      extractByHeader(record, "Alpha Omega Alpha (AOA) Status").equalsIgnoreCase("YES"),
      extractStepScore(extractByHeader(record, "USMLE Score (Step 1)")),
      extractStepScore(extractByHeader(record, "USMLE Score (Step 2)")),
      extractDate(extractByHeader(record, "Birth Date")),
      extractDate(extractByHeader(record, "Interview Date")),
      Seq()
    )

  private def extractLORData (record: CSVRecord) =
    LORMetaData(
      extractByHeader(record, "Letter of Recommendation File"),
      LORWriter(
        extractByHeader(record, "Letter Writer's Name"),
        extractByHeader(record, "Letter Writer's Academic Rank"),
        extractByHeader(record, "Anesthesiology Residency Program Director").equalsIgnoreCase("YES")
      )
    )

  private def extractByHeader (record: CSVRecord, header: String): String =
    record.get(header).toUpperCase().trim()

  private def extractStepScore (score: String): Option[Int] = {
    Try(score.toInt).toOption
  }

  private def extractDate (date: String): Option[Date] = {
    Try(dateFormatter.parse(date)).toOption
  }

  def fileNameToParticipantLetterId(fileName: String): String = fileName.split("[.]").head
  def fileNameToParticipantId(fileName: String): Int = fileNameToParticipantLetterId(fileName).replaceAll("[a-zA-Z]", "").toInt
}

File: analysis/lib/key_parser.py
import csv

key_path = '/home/mforkin/LOR/data/key.csv'


def parse_key():
    with open(key_path, 'r') as f:
        reader = csv.reader(f, delimiter=',')
        lines = [line for line in reader]
        headers_raw = enumerate([header.lower() for header in lines[0]])
        headers_map = {}
        for i, h in headers_raw:
            headers_map[h] = i
        metadata = {}
        [handle_line(line, metadata, headers_map) for line in lines[1:]]
        return metadata


def handle_line(line, metadata, header_map):
    id = line[0]
    user_data = {}
    if id in metadata:
        user_data = metadata[id]
    else:
        user_data = extract_metadata(line, header_map)
    metadata[id] = user_data
    lor_data = extract_lor_data(line, header_map)
    metadata[id]['lor_metadata'].append(lor_data)


def extract_metadata(line, header_map):
    return {
        'participant_number': line[header_map['participant number']],
        'medical_school': line[header_map['medical school']],
        'gender': line[header_map['gender']],
        'race': line[header_map['race (self identification)']],
        'aoa': line[header_map['alpha omega alpha (aoa) status']],
        'step1': line[header_map['usmle score (step 1)']],
        'step2': line[header_map['usmle score (step 2)']],
        'birth_date': line[header_map['birth date']],
        'interview_data': line[header_map['interview date']],
        'lor_metadata': []
    }


def extract_lor_data(line, header_map):
    return {
        'file_name': line[header_map['letter of recommendation file']],
        'writer_metadata': {
            'name': line[header_map["letter writer's name"]],
            'rank': line[header_map["letter writer's academic rank"]],
            'is_program_director': line[header_map['anesthesiology residency program director']] == 'yes',
        }
    }


key = parse_key()
